ocaml/idl/datamodel_types.ml

```diff
@@ -444,15 +444,123 @@ type release = {
 }
 [@@deriving rpc]
 
-type lifecycle_change =
-  | Prototyped
-  | Published
-  | Extended
-  | Changed
-  | Deprecated
-  | Removed
+module Lifecycle = struct
+  type state =
+    | Unreleased_s
+    | Prototyped_s
+    | Published_s
+    | Deprecated_s
+    | Removed_s
+  [@@deriving rpc]
+
+  type change =
+    | Prototyped
+    | Published
+    | Extended
+    | Changed
+    | Deprecated
+    | Removed
+  [@@deriving rpc]
+
+  type transition = change * string * string [@@deriving rpc]
+
+  type t = {state: state; transitions: transition list} [@@deriving rpc]
+
+  exception Invalid of string
+
+  type ('state, 'change) automaton = {
+      initial: 'state
+    ; next: 'change -> 'state -> 'state
+  }
+
+  let raise_invalid msg = raise (Invalid msg)
+
+  let string_of_state s = Rpc.string_of_rpc (rpc_of_state s)
+
+  let string_of_change c = Rpc.string_of_rpc (rpc_of_change c)
 
-and lifecycle_transition = lifecycle_change * string * string [@@deriving rpc]
+  let raise_invalid_next ~from ~into =
+    raise_invalid
+      (Printf.sprintf "Invalid transition %s from %s" (string_of_change into)
+         (string_of_state from)
+      )
+
+  (* lifecycle automaton, rules:
+     - Unreleased is the initial state
+     - Extended and Changed do not change the state
+     - Objects cannot be Changed nor Extended before being prototyped or after
+       being deprecated
+     - The rest of the changes are not idempotent, they cannot be applied twice
+       in a row
+     - Objects can only be removed when are prototyped or deprecated
+  *)
+  let automaton =
```

Comment:
Could this be simplified by using wildcards? There are only a few legal transitions in each state - express those explicitly and match all remaining ones with a wildcard and make them illegal?